### DISCUSSION TITLE
Использование диапазона значений на веб странице

### DISCUSSION THREAD

**Discussion by Bedmit** [Chimildic/goofy#92] [Помощь с алгоритмом] [answered]
Спасибо за добавления этой возможности, давно хотел попросить реализацию чего-то подобного.
Но во-первых не до конца понятно, что пишется в коде скрипта для ее использования.  Неплохо было бы в описании добавить в главе "Обработка" пример и для нее.

А во-вторых, можно ли расширить количество функций, где это можно использовать?
Я так понимаю, что сейчас это только `getRecomTracks`. А можно сделать подобное с функцией `mixin`?
Например в `Combiner.mixin(tracks, savedTracks, 1, 1, true)` среднее значение ползунка - соотношение 1 к 1. При сдвиге влево/вправо соответственно соотношение меняется.
И возможно ли сделать, что бы последнее выбранное значение диапазона запоминалось и выполнялось при автоматическом запуске скрипта, до тех пор пока оно не будет изменено?

**Comment by Chimildic** [comment]
> что пишется в коде скрипта для ее использования

Обработка ничем не отличается от остальных. Функция принимает `values` и в `range` содержит ключи `key` результатов с выбранными значениями.
```js
function test(values){
    console.log(values.range.tempo);
}
```

> расширить количество функций, где это можно использовать

Не совсем понимаю. Где использовать - никак не ограничено. Просто наиболее простой вариант с рекомендациями приходит первым. Поскольку там большое количество чисел вспоминается. 
В случае с соотношением, ничего не мешает сделать два ползунка. Или выстроить математику подсчета от одного. 

> последнее выбранное значение диапазона запоминалось

В простом случае, можно по умолчанию `default` поставить ноль или минус один. В обработке проверять значение. Если ноль, читать последнее из кэша, иначе брать вновь пришедшее. 
Усложненный вариант, сделать две обработки _get/set_ в коде, которые бы читали/писали кэш. И соответствующие правки кода на самой странице, чтобы их вызывать.

**Comment by Bedmit** [comment]
Не хотелось бы чтобы показалось что привередничаю, но ни чего из вышесказанного, меня не приблизило к пониманию ))
С моим опытом в программировании использование чего-либо в Goofy возможно только методом копирования конкретного примера, или куска кода из скрипта, с заменой каких-то значений под себя. Но просто из логики рассуждений, сам составить код увы еще не могу. Боюсь это проблема большинства пытающихся осилить Goofy без минимальной базы знаний в этой области - пока не разжевано до уровня "берем вот эти и вставляем сюда".

**Comment by Chimildic** [comment, marked as answer]
> что пишется в коде скрипта для ее использования

> Например в Combiner.mixin
```js
// launch.html
{
    name: 'mixinExample',
    text: 'Пример соотношения mixin',
    args: [
        {
            range: [
                { key: 'xRow', name: 'Подряд первый источник', min: 1, max: 10, step: 1, default: 1 },
                { key: 'yRow', name: 'Подряд второй источник', min: 1, max: 10, step: 1, default: 1 },
            ],
        },
    ],
}
```
```js
// main.gs
function mixinExample(values){
    let playlistTracks = Source.getTracks(playlistArray);
    let savedTracks = Source.getSavedTracks();
    let resultArray = Combiner.mixin(playlistTracks, savedTracks, values.range.xRow, values.range.yRow, true);
}
```

> последнее выбранное значение диапазона запоминалось
```js
// launch.html
{
    name: 'mixinExample',
    text: 'Пример соотношения mixin',
    args: [
        {
            // min и default в ноль
            range: [
                { key: 'xRow', name: 'Подряд первый источник', min: 0, max: 10, step: 1, default: 0 },
                { key: 'yRow', name: 'Подряд второй источник', min: 0, max: 10, step: 1, default: 0 },
            ],
        },
    ],
}
```
```js
// main.gs
function mixinExample(values){
    const filename = 'RangeValueHolder';
    let cached = Cache.read(filename);
    let xRow = values.range.xRow || cached.xRow || 1;
    let yRow = values.range.yRow || cached.yRow || 1;
    Cache.write(filename, { xRow: xRow, yRow: yRow });

    let playlistTracks = Source.getTracks(playlistArray);
    let savedTracks = Source.getSavedTracks();
    let resultArray = Combiner.mixin(playlistTracks, savedTracks, xRow, yRow, true);
}
```

**Reply by Bedmit** [reply]
Спасибо. Теперь даже понял, как по аналогии с этим применить в других функциях.

**Reply by Chimildic** [reply]
Что из этого оказалось в новинку? Кроме логики "запоминания"

**Reply by Bedmit** [reply]
Как выглядит то, что подставляется в функцию вместо цифровых значений -  `values.range.xRow` и `values.range.yRow`
